kotlin-native/runtime/src/main/kotlin/kotlin/native/Annotations.kt

```diff
@@ -83,3 +83,18 @@ public annotation class EagerInitialization
 @Retention(AnnotationRetention.BINARY)
 public actual annotation class CName(actual val externName: String = "", actual val shortName: String = "")
 
+/**
+ * Instructs the Kotlin compiler to remove this property or function from the public Objective-C API.
+ */
+@Target(AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.PROPERTY, AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+public actual annotation class ObjCRefined
```

Comment:
Could you please provide more details on why you selected this name for the annotation? What exactly does "refined" mean here?

Reply:
Apple has the [`NS_REFINED_FOR_SWIFT`](https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/improving_objective-c_api_declarations_for_swift) macro that hides the definition from Swift such that a more Swift friendly API can be defined. The `@SwiftRefined` annotation is a direct mapping to this macro (the `swift_private` attribute), hence the name. The `@ObjCRefined` annotation has a similar purpose, although slightly different since it completely removes the declaration from the ObjC API.

I think the following definition of "refined" is the most accurate in this case:
> with impurities or unwanted elements having been removed by processing.

Reply:
> Apple has the [`NS_REFINED_FOR_SWIFT`](https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/improving_objective-c_api_declarations_for_swift) macro that hides the definition from Swift such that a more Swift friendly API can be defined. The `@SwiftRefined` annotation is a direct mapping to this macro (the `swift_private` attribute), hence the name.

Maybe it should be `@RefinedForSwift`, `NSRefinedForSwift` or `@SwiftPrivate` then?

> The `@ObjCRefined` annotation has a similar purpose, although slightly different since it completely removes the declaration from the ObjC API.
>
> I think the following definition of "refined" is the most accurate in this case:
> > with impurities or unwanted elements having been removed by processing.

It doesn't seem like the original Swift name ("refined for Swift") is based on this definition, because nothing is actually removed. Also, the documentation says
> Applying the NS_REFINED_FOR_SWIFT macro exposes the existing Objective-C API for reuse in your refined API.

`@ObjCRefined` annotation doesn't in fact make anything for refining this API, because the intention is to hide this API and make another Kotlin declaration that will be exported. So it is more like `@HideFromObjC` or something.

I don't propose using exactly these names, but I think there is room for improvement in the current ones, to make them easier to understand and closer to analogues.

Reply:
Definitely! Some other possible names could be `@RefinedInSwift` and `@ObjCInternal`.

Reply:
I have renamed the annotations to `RefinesInSwift`, `RefinedInSwift`, `RefinesForObjC` and `RefinedForObjC`.

The meta-annotations are named `Refines` since the job of custom annotations is to refine the annotated declarations, the built-in annotations are named `Refined` indicating that the annotated declaration is already refined (by the developer).

For the Swift annotations `In` is used which indicates that the declaration is still exported since a refinement is declared in Swift, for the ObjC annotations `For` is used which indicates that the declarations aren't refined in ObjC but in Kotlin instead.